WIP: MSL: Support input/output blocks containing nested struct arrays - squashed

billhollings opened the pull request:
Replaces PR #1866, per initial review of that PR requesting commit squashing and syntax formatting.

Fixes numerous CTS tests of types `dEQP-VK.pipeline.interface_matching.vector_length.member_of_*`, passing complex nested structs between stages as stage I/O.

- Make `add_composite_member_variable_to_interface_block()` recursive to allow struct members to contain nested structs, building up member names and access chains recursively, and only add the resulting flattened leaf members to the synthetic input and output interface blocks.
- Recursively generate individual location numbers for the flattened members of the input/output block.
- Replace `to_qualified_member_name()` with `append_member_name()`.
- Update `add_variable_to_interface_block()` to support arrays as struct members, adding a member to input and output interface blocks for each element of the array.
- Pass name qualifiers to `add_plain_member_variable_to_interface_block()` to allow struct members to be arrays of structs, building up member names and access chains, and adding multiple distinct flattened leaf members to the synthetic input and output interface blocks.
- Generate individual location numbers for the individual array members of the input/output block.
- `SPIRVCrossDecorationInterfaceMemberIndex` references the index of a member of a variable that is a struct type. The value is relative to the variable, and for structs nested within that top-level struct, the index value needs to take into consideration the members within those nested structs.
- Pass `var_mbr_idx` to `add_plain_member_variable_to_interface_block()` and `add_composite_member_variable_to_interface_block()`, start at zero for each variable, and increment for each member or nested member within that variable.
- Add unit test `shaders-msl/vert/out-block-with-nested-struct-array.vert`
- Add unit test `shaders-msl/vert/out-block-with-struct-array.vert`
- Add unit test `shaders-msl/tese/in-block-with-nested-struct.tese`

Fix #1829.
Fix #1830.

Further work may be forthcoming as other CTS failures involving complex structs and arrays passed between shader stages are investigated.

@cdavis5e @HansKristian-Work Please review.

HansKristian-Work commented on spirv_msl.cpp:
```diff
@@ -3114,67 +3147,98 @@ void CompilerMSL::add_variable_to_interface_block(StorageClass storage, const st
 		{
 			bool masked_block = false;
+			uint32_t location = 0;
+			uint32_t var_mbr_idx = 0;
```
This is a bug. E.g. this shader emits wrong code:

```
#version 450

struct Foo
{
	float a;
	float b;
};

struct Bar
{
	vec2 a;
	vec2 b;
};

layout(location = 1) in Foo foos[4];
layout(location = 10) in Foo bars[4];
layout(location = 0) out vec4 FragColor;

void main()
{
	FragColor.x = foos[0].a;
	FragColor.y = foos[1].b;
	FragColor.z = foos[2].a;
	FragColor.w = bars[3].b.x;
}
```

HansKristian-Work replied:
I assume var_mbr_idx needs to start at size of ib_type.members.

HansKristian-Work replied:
The codegen is:
```
    bars[0].a = in.Foo_0_a; <-- alias with foos[] ?
    bars[0].b = in.Foo_0_b;
    bars[1].a = in.Foo_1_a;
    bars[1].b = in.Foo_1_b;
    bars[2].a = in.Foo_2_a;
    bars[2].b = in.Foo_2_b;
    bars[3].a = in.Foo_3_a;
    bars[3].b = in.Foo_3_b;
    out.FragColor.x = foos[0].a;
    out.FragColor.y = foos[1].b;
    out.FragColor.z = foos[2].a;
    out.FragColor.w = bars[3].b;
    return out;
```

billhollings replied:
The problem here is not `var_mbr_idx`, which marks the index into the var struct, not the ib_type struct. 

The core of the problem is that SPIRV-Cross uses **_type_** name to name the ib_type member, so if two vars use the same type, and they both get flattened into the same ib_type, we end up with a name conflict. This is resolved rather arbitrarily in the existing `add_member_name()` to append an arbitrary index when the struct is emitted. Unfortunately, this is too late for the fix-up lambdas that are created to do all the assignment copying above.

I noticed this issue while working on this, and I have a one-line fix in the new code that uses the **_var_** name, instead of the **_var type_** name to derive the ib_type member names. I did not include that change in this PR, because it changes the content of a whole ton of existing reference shaders, which I felt created a lot of review noise for this PR.

With that change in place, the resulting MSL shader becomes that as shown below.

I recommend we pull this PR in, since it's not breaking anything existing, and then I'll do a separate PR with that one-line change, changing all those existing reference shaders, and I'll include your shader above to demonstrate and test for that issue. I recommend doing it that way, because to do it in the existing code would be a larger change (because the existing `to_qualified_member_name()` does not pass in the var).

--------------

```
struct main0_in
{
    float foos_0_a [[user(locn1)]];
    float foos_0_b [[user(locn2)]];
    float foos_1_a [[user(locn3)]];
    float foos_1_b [[user(locn4)]];
    float foos_2_a [[user(locn5)]];
    float foos_2_b [[user(locn6)]];
    float foos_3_a [[user(locn7)]];
    float foos_3_b [[user(locn8)]];
    float bars_0_a [[user(locn10)]];
    float bars_0_b [[user(locn11)]];
    float bars_1_a [[user(locn12)]];
    float bars_1_b [[user(locn13)]];
    float bars_2_a [[user(locn14)]];
    float bars_2_b [[user(locn15)]];
    float bars_3_a [[user(locn16)]];
    float bars_3_b [[user(locn17)]];
};

fragment main0_out main0(main0_in in [[stage_in]])
{
    main0_out out = {};
    spvUnsafeArray<Foo, 4> foos = {};
    spvUnsafeArray<Foo, 4> bars = {};
    foos[0].a = in.foos_0_a;
    foos[0].b = in.foos_0_b;
    foos[1].a = in.foos_1_a;
    foos[1].b = in.foos_1_b;
    foos[2].a = in.foos_2_a;
    foos[2].b = in.foos_2_b;
    foos[3].a = in.foos_3_a;
    foos[3].b = in.foos_3_b;
    bars[0].a = in.bars_0_a;
    bars[0].b = in.bars_0_b;
    bars[1].a = in.bars_1_a;
    bars[1].b = in.bars_1_b;
    bars[2].a = in.bars_2_a;
    bars[2].b = in.bars_2_b;
    bars[3].a = in.bars_3_a;
    bars[3].b = in.bars_3_b;
    out.FragColor.x = foos[0].a;
    out.FragColor.y = foos[1].b;
    out.FragColor.z = foos[2].a;
    out.FragColor.w = bars[3].b;
    return out;
}
```

HansKristian-Work replied:
Right. Makes more sense to flatten based on the variable name I guess. Only annoyance is when the SPIRVariable does not have a name, but the type does.

HansKristian-Work commented on spirv_msl.cpp:
```diff
@@ -3114,67 +3147,98 @@ void CompilerMSL::add_variable_to_interface_block(StorageClass storage, const st
 		else
 		{
 			bool masked_block = false;
+			uint32_t location = 0;
```
The use of location here is kinda confusing. It's used as a sentinel *and* a valid value at the same time. I would suggest using a more clear sentinel like UINT32_MAX or something.

billhollings replied:
Fixed.

HansKristian-Work commented on spirv_msl.cpp:
```diff
@@ -3114,67 +3147,98 @@ void CompilerMSL::add_variable_to_interface_block(StorageClass storage, const st
-					bool attribute_load_store =
-					    storage == StorageClassInput && get_execution_model() != ExecutionModelFragment;
-					bool storage_is_stage_io = variable_storage_requires_stage_io(storage);
+						location++;		// Skip this location
```
What if the type consumes more than one location?

billhollings replied:
Fixed. Now letting location be re-resolved on next unmasked var member.

HansKristian-Work commented on spirv_msl.cpp:
```diff
@@ -3114,67 +3147,98 @@ void CompilerMSL::add_variable_to_interface_block(StorageClass storage, const st
+						if (builtin == BuiltInClipDistance || builtin == BuiltInCullDistance)
+							is_builtin = false;
+
+						string mbr_name_qual = to_name(var_type.self) + (elem_cnt == 1 ? "" : join("_", elem_idx));
```
Should just use join() here.

billhollings replied:
Fixed and simplified.

HansKristian-Work commented on spirv_msl.cpp:
```diff
@@ -12982,8 +13046,8 @@ string CompilerMSL::to_name(uint32_t id, bool allow_alias) const
-// Returns a name that combines the name of the struct with the name of the member, except for Builtins
-string CompilerMSL::to_qualified_member_name(const SPIRType &type, uint32_t index)
+// Appends the name of the member to the variable qualifier string, except for Builtins.
+string CompilerMSL::append_member_name(const string qualifier, const SPIRType &type, uint32_t index)
```
`const string` should probably be `const string &` here.

billhollings replied:
Fixed.

HansKristian-Work commented on spirv_msl.cpp:
```diff
@@ -2754,48 +2781,54 @@ void CompilerMSL::add_plain_member_variable_to_interface_block(StorageClass stor
-		mark_location_as_used_by_shader(locn, get<SPIRType>(mbr_type_id), storage);
+		set_member_decoration(ib_type.self, ib_mbr_idx, DecorationLocation, location);
+		mark_location_as_used_by_shader(location, get<SPIRType>(mbr_type_id), storage);
+		location++;
```
What if type consumes more than one location?

billhollings replied:
At this point, we've tunnelled into structs, arrays, and matrices. I think we should be down to a single location by now.

HansKristian-Work replied:
Should probably just add a comment about that, then it's okay I guess.

billhollings replied:
Comment added.

HansKristian-Work commented on spirv_msl.cpp:
```diff
@@ -2572,6 +2575,26 @@ void CompilerMSL::add_composite_member_variable_to_interface_block(StorageClass
+																 var, mbr_type, sub_mbr_idx,
+																 meta, mbr_name, var_chain,
+																 location, var_mbr_idx);
+				var_mbr_idx++;
```
With recursion, cannot add_composite_member_variable_to_interface_block increment more than once? Maybe var_mbr_idx should be mutable and incremented by the add_composite_member_variable function itself?

billhollings replied:
Yes. Recursive incrementation happens here, because at this point we are inside `add_composite_member_variable_to_interface_block()`, and calling it for the next level.

Could move the incrementing into `add_composite_member_variable_to_interface_block()`, but would also have to also move it into `add_plain_member_variable_to_interface_block()`, and still handle the masking that doesn't invoke either.

HansKristian-Work replied:
I still don't understand the purpose this exact line. I tried a test shader like this:

```
#version 450
layout(quads) in;

layout(location = 0) out vec4 FragColor;

struct Bar
{
	float c;
	float d;
};

struct Foo
{
	Bar a;
	Bar b;
};

layout(location = 0) in Foo foos[];

void main()
{
	FragColor.x = foos[0].a.c;
	FragColor.y = foos[1].a.d;
	FragColor.z = foos[0].b.c;
	FragColor.w = foos[1].b.d;
}
```
and this ends up emitting code like:
```
#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct Bar
{
    float c;
    float d;
};

struct Foo
{
    Bar a;
    Bar b;
};

struct main0_out
{
    float4 FragColor [[user(locn0)]];
    float4 gl_Position [[position]];
};

struct main0_in
{
    float Foo_a_c [[attribute(0)]];
    float Foo_a_d [[attribute(1)]];
    float Foo_b_c [[attribute(2)]];
    float Foo_b_d [[attribute(3)]];
};

struct main0_patchIn
{
    patch_control_point<main0_in> gl_in;
};

[[ patch(quad, 0) ]] vertex main0_out main0(main0_patchIn patchIn [[stage_in]])
{
    main0_out out = {};
    out.FragColor.x = patchIn.gl_in[0].Foo_a_c;
    out.FragColor.y = patchIn.gl_in[1].Foo_a_d;
    out.FragColor.z = patchIn.gl_in[0].Foo_a_d; <-- Misaligned?
    out.FragColor.w = patchIn.gl_in[1].Foo_b_c;
    return out;
}
```
If I understand it correctly, var_mbr_idx is supposed to uniquely define a member inside the composite object, but something seems off, e.g.:

```
		set_extended_member_decoration(ib_type.self, ib_mbr_idx, SPIRVCrossDecorationInterfaceMemberIndex, var_mbr_idx);
```
Here var_mbr_idx takes 0, 1, 3, 4. 2 is skipped due to this line incrementing things additionally.

billhollings replied:
Hmmm...after digging into this more...the problem is not really how `var_mbr_idx` is calculated. Moving it inside `add_composite_member_variable_to_interface_block()` and `add_plain_member_variable_to_interface_block()` doesn't help this use case, and it in turn breaks `shaders-msl/asm/frag/pull-model-interpolation.asm.msl23.frag` and other test shaders.

The problem for this use case is that `emit_tessellation_access_chain()`, `emit_tessellation_io_load()`, `fix_up_interpolant_access_chain()`, and `fix_up_interface_member_indices()` all need to handle the recursion as well.

There is so much existing complex and repetitive code surrounding this, particularly around tessellation handling, that this is going to get really ugly. I did some work on this earlier, but pulled back to the current PR state because of that. 

But to do this right, covering all recursive structures is going to take some significant rework. That's all too much to take care of right now. I'm going to move this PR to WIP for now. I'll try to find some time to properly address this soon.

HansKristian-Work replied:
Sure, as long as the code is sane and doesn't cause regressions we can merge.

HansKristian-Work commented:
I meant that formatting for the PR itself should be fixed. Having 3ksloc of all module clang-format should be done after in a separate PR if anything.

billhollings commented:
> I meant that formatting for the PR itself should be fixed. Having 3ksloc of all module clang-format should be done after in a separate PR if anything, since it's not reviewable to do it like this.

Yeah. Xcode uses different indent rules. Figured it was simpler to just do it automatically than try to edit by hand.

HansKristian-Work commented on shaders-msl/vert/out-block-with-struct-array.vert:
```diff
@@ -0,0 +1,24 @@
+void main()
+{
+    gl_Position = _17;
+    _25[2]._m1 = vec4(-4.0, -9.0, 3.0, 7.0);
```
The tests should not use reserved identifiers, makes the test output very ugly to read.

billhollings replied:
Fixed.

HansKristian-Work reviewed (changes requested):
Found bug. Also, the clang-format commit should be rebased out of this PR for now.

billhollings commented:
> Also, the clang-format commit should be rebased out of this PR for now.

I've reverted the clang-format, and rebased all other review changes into a single commit.

HansKristian-Work commented:
> > Also, the clang-format commit should be rebased out of this PR for now.
> 
> I've reverted the clang-format, and rebased all other review changes into a single commit.

Was this pushed? Don't see any changes.

billhollings commented:
> Was this pushed? Don't see any changes.

Sorry. Still working through other items and answering as I go. 

Will rebase and push all soon. I'll re-request a review once that done.
